Storage Access API should not be usable in insecure contexts

charliewolfe opened the pull request:
#### d4714a11767be39e4a82f90269711c728cf986fa
<pre>
Storage Access API should not be usable in insecure contexts
<a href="https://bugs.webkit.org/show_bug.cgi?id=296835">https://bugs.webkit.org/show_bug.cgi?id=296835</a>
<a href="https://rdar.apple.com/157337423">rdar://157337423</a>

Reviewed by Tim Nguyen.

Implements 3.2.5 in <a href="https://privacycg.github.io/storage-access/#the-document-object">https://privacycg.github.io/storage-access/#the-document-object</a>

This progresses both of the tests below, but we won’t have CI coverage because we lack a local DNS
resolver.

imported/w3c/web-platform-tests/storage-access-api/hasStorageAccess-insecure.sub.window.html
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-insecure.sub.window.html

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, StorageAccessPromptSiteWithTrigger)):
(TEST(ResourceLoadStatistics, StorageAccessSupportMultipleSubFrameDomains)):
(TEST(ResourceLoadStatistics, StorageAccessGrantMultipleSubFrameDomains)):
Update these tests to work with an HTTPS server.

* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::hasStorageAccessQuickCheck):
(WebCore::DocumentStorageAccess::requestStorageAccess):

Canonical link: <a href="https://commits.webkit.org/298179@main">https://commits.webkit.org/298179@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79c4f08df44cfef8a2f675114aa7be0714e4bd73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120712 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65273 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42854 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87061 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41971 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27828 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67452 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20996 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64394 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97206 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21109 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123919 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31021 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95881 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41938 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99059 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95664 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40828 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18665 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37620 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18344 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41440 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46950 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41026 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44333 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42776 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/DocumentStorageAccess.cpp:
```diff
@@ -119,6 +119,11 @@ void DocumentStorageAccess::hasStorageAccess(Ref<DeferredPromise>&& promise)
+    if (!document->isSecureContext()) {
+        promise->resolve<IDLBoolean>(false);
+        return;
+    }
```
you can move this check into `hasStorageAccessQuickCheck`

nt1m commented on Source/WebCore/dom/DocumentStorageAccess.cpp:
```diff
@@ -198,6 +203,11 @@ void DocumentStorageAccess::requestStorageAccess(Ref<DeferredPromise>&& promise)
         return;
     }
 
+    if (!document->isSecureContext()) {
+        promise->reject(ExceptionCode::NotAllowedError);
+        return;
+    }
```
Same here for `requestStorageAccessQuickCheck` (although maybe it won't reject with the same exception than the spec???, please double check)

charliewolfe replied:
I can't here. I need the exception the spec specifies.

charliewolfe committed:
Storage Access API should not be usable in insecure contexts
https://bugs.webkit.org/show_bug.cgi?id=296835
rdar://157337423

Reviewed by Tim Nguyen.

Implements 3.2.5 in https://privacycg.github.io/storage-access/#the-document-object

This progresses both of the tests below, but we won’t have CI coverage because we lack a local DNS
resolver.

imported/w3c/web-platform-tests/storage-access-api/hasStorageAccess-insecure.sub.window.html
imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-insecure.sub.window.html

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, StorageAccessPromptSiteWithTrigger)):
(TEST(ResourceLoadStatistics, StorageAccessSupportMultipleSubFrameDomains)):
(TEST(ResourceLoadStatistics, StorageAccessGrantMultipleSubFrameDomains)):
Update these tests to work with an HTTPS server.

* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::hasStorageAccessQuickCheck):
(WebCore::DocumentStorageAccess::requestStorageAccess):

Canonical link: https://commits.webkit.org/298179@main
